MessageSigning/MessageSigning.md

```diff
@@ -0,0 +1,433 @@
+
+```
+COSE_Encrypt0 = [
+    Headers,
+    ciphertext : bstr / nil,
+]
+
+PasswordEncryption = 16 (COSE_Encrypt0)
+```
+
+The `Headers` MUST have `enc_type: 1` in the `unprotected` field.
+
+The `ciphertext` is simply the result of [the Yoroi encryption spec](https://github.com/Emurgo/yoroi-frontend/blob/737595fec5a89409aacef827d356c9a1605515c0/docs/specs/code/ENCRYPT.md) on the `signed_message`.
+
+##### Public key based encryption
+
+We only allow encrypting based on `ED25519` public keys (the ones used for Cardano). To encrypt based on these public keys, you must
+
+1) Compute a password consisting of 22 case-sensitive alphanumeric (a–z, A–Z, 0–9) characters (this gives you ~128 bits of entropy)
+
+Now, for each receiver, you must
+
+2) Compute an ephemeral key pair using `ED25519 extended`
+1) Compute the shared DH secret between the private key from step (1) and the public key received (using the `exchange` functionality)
+1) Compute the `Blake2b512` hash of the secret
+1) Use this as the password to encrypt the password in (1) using [the Yoroi encryption spec](https://github.com/Emurgo/yoroi-frontend/blob/737595fec5a89409aacef827d356c9a1605515c0/docs/specs/code/ENCRYPT.md)
+
+The structure will look like the following:
+
+```
+COSE_Encrypt = [
+  Headers,
+  ciphertext : bstr / nil, ; contained signed_message encrypted with random password
+  recipients : [+COSE_recipient]
+]
+
+COSE_recipient = [
+  Headers,
+  ciphertext : bstr / nil, ; contains random password encrypted with shared secret
+]
+
+PubKeyEncryption = 96 (COSE_Encrypt)
+```
+
+The `Headers` for the body MUST have `enc_type: 2` in the `unprotected` field.
+
+The `Headers` for the recipient MUST have a `eph_pub: bstr` label containing the public key of the ephemeral keypair.
+
+#### Version
+
+The `Headers` for the body MUST have `version: uint` in the `unprotected` field. See Changelog section for possible versions.
+
+#### CWT
+
+CWT tags MAY be introduced inside `protected` headers as a CBOR map with the key `cwt`. It is RECOMMENDED to use these to avoid replay attacks
+
+#### Payload encoding
+
+To solve `E3`, `signed_message` body header MUST contain `hashed: bool` as an `unprotected` header which defines whether or not we signed the `payload` OR the `Blake2b224` hash of the `payload`. The hash MUST be used in the following two cases
+
+1) The size of the raw `payload` would otherwise be too big to fit in hardware wallet memory (see E1). Note that the exact size for which this is the case depends on the device.
+1) The payload characters (ex: non-ASCII) that cannot be displayed on the hardware wallet device (see E3)
+
+We RECOMMEND showing the user the full payload on the device if possible because it lowers the attack surface (otherwise the user has to trust that the hash of the payload was calculated correctly).
+
+`Blake2b224` was chosen specifically because `224` bits is already a long string for hardware wallets.
```

Comment:
If we allow signing the hash of the transaction directly by the hardware wallet without making the device hash the payload by itself I see the risk of the user unintentionally signing a valid transaction hash without them being aware. The user may even see the hash on the device but how does he know for sure it isn't a valid transaction hash?

A possible workaround could be appending a header to the hash to be signed so it does not have the same structure as a valid tx hash, thus making the attack (at least as I described it) impossible.

Reply:
indeed, digest length for tx hashes is 256, not 224

Reply:
When I wrote this spec, Cardano had a feature (in both Byron and Jormungandr) where you would sign objects with a specific prefix to avoid any chance of collision like this (ex: transactions would be signed as ('x' || payload), etc.) but this was removed in Haskell Shelley.
This message signing spec used to reserve a new prefix for signed messages, but since Cardano removed this prefix feature, I removed the prefix from this spec also. It's entirely possible passing just the hash to the hardware wallet no longer works after this change (and you would instead have to pass the full body and calculate the hash on the device itself)

Reply:
I reviewed this again and I think it's still fine as-is because you're signing the payload wrapped in a structure so the probability of a collision is basically 0.